personalrobotics/ada_feeding
# Add Node that kills itself when the watchdog fails

# Description

This PR does the following:

1. Separates the watchdog listener logic from `create_action_servers.py` to its own class.
2. Creates a node that kills itself if the watchdog fails. This node exposes a service that allows it to be toggled on and off.
    1. As part of that, I changed the watchdog listener behavior to set the default `watchdog_failed` to False. This means that if the watchdog is not publishing when the listener starts listening, it will take 0.5sec (configureable by a param) before the watchdog listener registers a failure (because that is how long we allow the watchdog to go without publishing). The reason for this change is so that if we create a watchdog listener and query it before a watchdog message has been received (e.g., if it is listening at a faster rate than the watchdog), it doesn't immediately think the watchdog failed.
3. Modifies `ada_planning_scene.py` to wait until MoveIt is running before publishing the messages to populate the planning scene. This is because of an ordering dependency: `ada_feeding_launch.xml` launches both the watchdog and `ada_planning_scene.py`. But the MoveIt launch requires the watchdog to be running before it is launched, and `ada_planning_scene.py` requires MoveIt to be running before it is launched. This change resolves the ordering issue.
4. Modifies the dummy F/T sensor to expose a re-taring service, that stops it from publishing messages for 0.75 sec (to mimic the fact that the actual F/T sensor stops publishing messages for that long when re-taring).

# Testing procedure

Pull the code from [ada_ros2#10](https://github.com/personalrobotics/ada_ros2/pull/10).

1. Verify that `create_action_servers.py` still works as expected:
    - [x] Run the dummy FT sensor (`ros2 run dummy_ft_sensor.py`), the ada_feeding launchfile, and MoveIt in RVIZ. Send a MoveAbovePlate action and verify that it succeeds.
    - [x] Kill the dummy FT sensor. Send a MoveAbovePlate action and verify that it is rejected.
    - [x] Re-start the dummy FT sensor. Send a MoveAbovePlate action, and kill the dummy FT sensor before the action completes. Verify that the action aborts.
3. Verify that re-taring the F/T sensor works properly. 
    - [x] Run the dummy F/T sensor, echo the F/T topic, and re-tare the sensor (`ros2 service call /wireless_ft/set_bias std_srvs/srv/SetBool "{data: true}"`). Verify that the messages stop publishing for 0.75 sec.

# Before opening a pull request

# Before Merging
